Improve performance and code style, refactor, partial rewrite

mforthewin opened the pull request:
The program in its current state takes a lot of time to execute compared to programs like `ps`.

The reasons for this behavior are for example:
* The usage of `nftw()`, which results in LOTS of unnecessary `newfstatat()` syscalls
* The reading of files one byte at a time
* Possibly also regex, as this is not needed for the simple format of  `"/proc/<pid>/stat"`

This becomes clear when profiling the syscalls:
```
$ strace -c zps -s
% time     seconds  usecs/call     calls    errors syscall
------ ----------- ----------- --------- --------- ----------------
 69.37    1.386260           4    333730           newfstatat
 10.63    0.212523           8     25761         1 getdents64
 10.41    0.208056           3     68370           read
  3.89    0.077740           5     14517      1171 openat
  3.39    0.067740           2     25748           fcntl
  2.12    0.042305           3     13346           close
  0.10    0.002090           2       705           write
  0.05    0.000917           8       107           brk
  0.01    0.000256          85         3           mprotect
  0.01    0.000139         139         1           munmap
  0.00    0.000098          12         8           mmap
  0.00    0.000039          39         1           clock_gettime
  0.00    0.000038          38         1           getrandom
  0.00    0.000035          35         1           prlimit64
  0.00    0.000034          34         1           rseq
  0.00    0.000021          10         2         1 arch_prctl
  0.00    0.000020          20         1           dup2
  0.00    0.000018          18         1           set_tid_address
  0.00    0.000018          18         1           set_robust_list
  0.00    0.000000           0         2           pread64
  0.00    0.000000           0         1         1 access
  0.00    0.000000           0         1           execve
------ ----------- ----------- --------- --------- ----------------
100.00    1.998347           4    482309      1174 total
```
Those are almost 500k syscalls! If we compare this to the new version of the code, the difference is huge:
```
$ strace -c build/zps -s
% time     seconds  usecs/call     calls    errors syscall
------ ----------- ----------- --------- --------- ----------------
 39.79    0.003438           9       358           read
 37.96    0.003280           9       362           openat
 22.07    0.001907           5       362           close
  0.13    0.000011           3         3           write
  0.05    0.000004           2         2           getdents64
  0.00    0.000000           0         8           mmap
  0.00    0.000000           0         3           mprotect
  0.00    0.000000           0         1           munmap
  0.00    0.000000           0         3           brk
  0.00    0.000000           0         1         1 ioctl
  0.00    0.000000           0         2           pread64
  0.00    0.000000           0         1         1 access
  0.00    0.000000           0         2           dup2
  0.00    0.000000           0         1           execve
  0.00    0.000000           0         2         1 arch_prctl
  0.00    0.000000           0         1           set_tid_address
  0.00    0.000000           0         4           newfstatat
  0.00    0.000000           0         1           set_robust_list
  0.00    0.000000           0         1           prlimit64
  0.00    0.000000           0         1           getrandom
  0.00    0.000000           0         1           rseq
------ ----------- ----------- --------- --------- ----------------
100.00    0.008640           7      1120         3 total
```

Also looking at real execution times, we can see the following speedup:
```
$ time zps -s

real	0m1.106s
user	0m0.320s
sys	0m0.743s
$ time build/zps -s

real	0m0.011s
user	0m0.003s
sys	0m0.008s
```

Another aspect is that having a `-f` flag along with `nftw()` is not sensible, as this could happen on a box with `2` GB of RAM:
```
[arch@archlinux ~]$ ./zps -f 111111111111111
PID     PPID    STATE               NAME COMMAND
Killed
[arch@archlinux ~]$ sudo dmesg | tail -n 5
[ 5856.621333] oom-kill:constraint=CONSTRAINT_NONE,nodemask=(null),cpuset=/,mems_allowed=0,global_oom,task_memcg=/user.slice/user-1000.slice/session-1.scope,task=zps,pid=24883,uid=1000
res=success'
[arch@archlinux ~]$ ./zps -f 11111111111111111
PID     PPID    STATE               NAME COMMAND
=================================================================
==24937==ERROR: AddressSanitizer: out of memory: allocator is trying to allocate 0x136b39e38 bytes
    #0 0x7fedfcee1359 in __interceptor_malloc /usr/src/debug/gcc/gcc/libsanitizer/asan/asan_malloc_linux.cpp:69
    #1 0x7fedfcd1fa6e  (/usr/lib/libc.so.6+0x101a6e) (BuildId: 8bfe03f6bf9b6a6e2591babd0bbc266837d8f658)

==24937==HINT: if you don't care about these errors you may set allocator_may_return_null=1
SUMMARY: AddressSanitizer: out-of-memory /usr/src/debug/gcc/gcc/libsanitizer/asan/asan_malloc_linux.cpp:69 in __interceptor_malloc
==24937==ABORTING
```
As we can see, it is trying to allocate more than `5.21` GB and the program is getting OOM-killed! This defeats the actual purpose of the program (freeing resources). Thus,  I propose these changes to fix any issues of performance or resource usage. Additionally, I addressed [TODO#2](https://github.com/orhun/zps#todos), and the program now sends a `SIGCHLD` signal by default (without the reap option). Also, as the `static` global array in the program was not optimal, a simple vector implementation is used now, eliminating the overflow issues from before. Please see commit b20d95452026bc17957bf13e5f19780657ed2477 for more specific details.

Commit:
Improve readability (uniformity, naming, includes)

Commit:
De-cluttering, new types, reduced usage of globals

Commit:
Introduce new enums, change format functions

Commit:
Change to realtime clock

Commit:
Improve performance and style, partial rewrite

This commit addresses TODO#2 by universalizing the functions handling
the zombie processes and sending `SIGCHLD` in the case where the
user did not explicitly tell the program to reap zombies (reflected
in the `terminate` member of our settings. Whenever a signal
is sent, the used signal is displayed to the user.

Some parts of old functions were put into new functions to also
improve readability.

The most substantial improvement lies in the performance aspect
of the program. The use of `nftw()` in the old code was the
biggest bottleneck, as an insane, unnecessary number of syscall
are issued for each process. Thus, the directory and file reading
part was adapted, using manual `readdir()` calls. As the `-f` flag
did not make sense to have in the program anyway (and could lead
to OOM kill), it was also dropped.

Regex was also dropped, as its use was unnecessary and complicated
things, and the parsing of `"/proc/<pid>/stat"` was rewritten.

As kernel processes were not excluded in the program before,
this commit checks for processes which should be left out
of our considerations (PID `2`, as well as its children;
`init` is also excluded).

Display widths for the output are now defined in the accompanying
header, and should be in line with the `proc(5)` documentation.

The code for the prompt was improved with respect to the handling
of invalid user input and coloring.

Finally, to completely remove the unnecessary globals, a vector
structure was defined in the header, which can dynamically grow.
As the old version of the code did not check for overflows in this
context, this is also a great improvement.

Commit:
Add newline at end of file in Dockerfile

Commit:
Adapt code in `zproc.c` to match `zps.{c,h}`

Commit:
Update manpage to match source

Commit:
Remove the `-f` flag test

Commit:
Remove unnecessary comments and update category

Commit:
Add missing newline at EOF in `CMakeLists.txt`

Commit:
Remove notions of `-f` from the READMEs

Commit:
Add a `.clang-format` file consistent with sources

Commit:
Remove the other `-f` test

Commit:
Add missing `<limits.h>` include

Commit:
Use custom signal abbreviation function

Commit:
Remove `stdlib.h.gcov` from `rm` command

Commit:
Change array to `static`

Commit:
Remove redundant print (will never happen (`proc(5)`)

Commit:
Add `const` qualifiers to be more clear

Commit:
Add even more `const` qualifiers

Commit:
Change function pre-conditions to `assert()` calls

orhun commented:
Oh wow, this is huge! `zps` was my first C project when I started college so it is good to see someone taking it apart years later 🐻 

I believe there are a lot of things that I will learn from this code so I will review it carefully. Obviously I'm not a C guru so expect some beginner questions :)

Anyhoo, thanks a lot for this. I just tested it and the speed difference is freaking unbelievable ⚡

orhun commented on src/zps.h:
```diff
@@ -18,22 +18,203 @@
+#include <sys/types.h>
+
+/* Version number string */
+#define VERSION "1.2.9"
```
Early review comment:
```suggestion
#define VERSION "2.0.0"
```

mforthewin committed:
Bump version

Co-authored-by: Orhun Parmaksız <orhunparmaksiz@gmail.com>

Commit:
Drop unused macro

Commit:
Define `NDEBUG` in `CMakeLists.txt` and `Makefile`

This will disable the pre-condition assertions in all functions,
which should never be false for releases anyway.

Commit:
Extend color functionality, refactor CLI options

This commit ...

* allows users to directly specify signals to
send to zombie parents (using common signal
numbers, abbreviations and abbreviations with
"sig" in front of them (case-insensitive)

* changes existing command-line options to simpler
ones

* places constraints on the possible combinations
of command-line options used

* adds functionality to detect if the program's
output is, for example, piped to another program
that will automatically disable ANSI color codes

* adds a `--no-color` option to manually disable
the default coloring

The manual page and READMEs would still need
to be updated.

Commit:
Update tested flags

mforthewin commented:
I've also made some changes to the CLI options, adding some new functionalities, such as a signal option. Let me know what you think – the man page and the READMEs would need an update before we should merge this, so I marked it as a draft again.

orhun commented:
Looks good! Let's go ahead and finalize this, I can provide a review soon.

Commit:
Update demos, READMEs and help texts

Commit:
Bump year

mforthewin commented:
> Looks good! Let's go ahead and finalize this, I can provide a review soon.

Alright! I've updated the remaining files and also added updated demo GIFs.

orhun commented on README.md:
```diff
@@ -119,41 +123,50 @@ Usage:
 
 ## TODO(s)
 
-- Improve listing processes for long process names.
-- Send `SIGCHLD` signal to the parent instead of terminating it.
+- Check Korean translation for mistranslations introduced in new release.
+- Maybe extend `cmdline`/`COMMAND` output to extend to the maximum line length,
+although this would still lead to varying results if one were to look for
+a specific string in the "interactive"/non-piped `zps` output manually
+compared to piping `stdout` to `grep`, as the maximum line length would
+not have the same limit. Thus, it might actually be desirable to actually
+limit the lengths in any case.
```
Can you remove this section from README.md and create separate issues for these? I think that would make them more visible for outside contributions.

mforthewin replied:
Yes, that's actually something I thought about – will do!

orhun commented on src/zps.c:
```diff
@@ -15,439 +15,797 @@
+
+#include "zps.h"
+
+static const char *const abbrevs[NSIG] = {
```
Can you add a one-line top comment explaining what this is? Also, would that be applicable to define these in the header file? (I don't have much experience with C/C++, just curious,)

mforthewin replied:
Thanks for pointing that out. Yes, it would be possible to define this in the header file, but usually we would declare to-be-allocated objects directly in the `.c` files – you would find the same patterns in `glibc` implementations or the Linux source code, for example.

orhun commented on src/zps.c:
```diff
@@ -15,439 +15,797 @@
+ */
+static int sig_str_to_num(const char *sig_str)
+{
+    assert(sig_str);
```
I've seen these `assert`s throughout the code I'm assuming these are for diagnostics, right?

mforthewin replied:
Yes, these runtime assertions are simply there for testing and debugging purposes, checking the preconditions of the functions which should hold at all times – for example, we do not intend to pass `NULL` pointers. In the releases, we can disable those runtime assertions at compile time by defining `NDEBUG`, as was added to the CMake configuration and the Makefile. The testing container should still have the assertions enabled, though.

orhun replied:
Got it! Do you think we should disable them for the container as well?

mforthewin replied:
We probably should do so in `docker/Dockerfile`, but not in `test.sh`. I will add the flag to the Dockerfile for this PR.

orhun commented on src/zps.c:
```diff
@@ -15,439 +15,797 @@
-#include <sys/stat.h>
-#include <string.h>
+#include <getopt.h>
+#include <limits.h>
```
My LSP says `limits.h` is not used anywhere, can you check?

mforthewin replied:
Well, this include had to be used because we're using `PATH_MAX` for this, as long as it's defined in `<limits.h>`. If we were to use different filesystems from the `procfs`, we should rather not be doing it like this, as, actually, longer paths could still be used in theory. A probably more portable version for non-`procfs` purposes would be using `pathconf(3)` with `_PC_PATH_MAX`, but then we would need dynamic allocation with `malloc` or use funky stuff like variable-length arrays (not recommended).

mforthewin replied:
And also, the build with `musl` on the Alpine container failed because of the previously missing include.

orhun replied:
I see, then there is something funky going on with my LSP.

Commit:
Improve code clarity and remove TODOs (issues TBA)

Commit:
Disable runtime assertions for the `Dockerfile`

mforthewin commented:
Is there anything left to do? Personally, I think this is in a good state now for merging.

orhun commented:
Yes, this is awesome, thanks a lot once again!

I will create a release now and share it on my socials tomorrow.
